src/Pact/Runtime/Capabilities.hs

```diff
@@ -51,19 +49,27 @@ type ApplyMgrFun e = Def Ref -> [PactValue] -> [PactValue] -> Eval e (Either Pac
 noopApplyMgrFun :: ApplyMgrFun e
 noopApplyMgrFun _ mgd _ = return $ Right mgd
 
-
+-- | Get any cap that is currently granted, of any scope.
 grantedCaps :: Eval e (S.Set Capability)
-grantedCaps = S.fromList . toList <$> use evalCapabilities
+grantedCaps = toSet <$> getAllStackCaps <*> getAllManaged
+  where
+    toSet a b = S.fromList $ a ++ b
+    getAllManaged = concatMap toList . S.toList <$> use (evalCapabilities . capManaged)
 
 -- | Matches Managed -> managed list, callstack and composed to stack list
 -- Composed should possibly check both ...
 capabilityGranted :: CapScope m -> Capability -> Eval e Bool
-capabilityGranted scope cap = memSet <$> scopeCaps
+capabilityGranted scope cap = elem cap <$> scopeCaps
   where
-    memSet = S.member cap . S.fromList . concatMap toList
+    scopeCaps :: Eval e [Capability]
     scopeCaps = case scope of
-      CapManaged _ -> use $ evalCapabilities . capManaged
-      _ -> use $ evalCapabilities . capStack
+      -- Managed: only check top-level, not composed
+      CapManaged _ -> map _csCap . S.toList <$> use (evalCapabilities . capManaged)
```

Comment:
toList are unnecessary here - `Set` is foldable, so implements `elem`. We could just do 

```haskell
S.map _csCap <$> use (evalCapabilities . capManaged)
```

then 
```haskell
S.elem cap <$> scopeCaps
```

Reply:
kk